### PR TITLE
Disable 'use local function' in a few important language cases.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/23118
Fixes https://github.com/dotnet/roslyn/issues/22672

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @Pilchie .

**Comment by CyrusNajmabadi** [comment]
Note: this fixes all of the cases from https://github.com/dotnet/roslyn/issues/22672 except for this one:

```c#
using System;

class Enclosing<T>
{
    delegate T MyDelegate(T t);
    static void Callee(MyDelegate d) => d(default);

    public class Class<T>
    {
        public void Caller()
        {
            MyDelegate local = x => x;
            Callee(local);
        }
    }
}
```

We will still offer the feature here, which will produce the broken code:

```c#
using System;

class Enclosing<T>
{
    delegate T MyDelegate(T t);
    static void Callee(MyDelegate d) => d(default);

    public class Class<T>
    {
        public void Caller()
        {
            T local(T x) => x;
            Callee(local);
        }
    }
}
```

The reason this is broken is that 'local' is now reference hte 'T' type parameter from the containing class ```Class<T>```, whereas before it was using hte 'T' type parameter from ```Enclosing<T>```.  This breaks because, of course, there's no language facility to reference a specific type parameter with the same name (unlike with types, when you can do things like global::)

The reason i didn't fix this was that it seemed like a pretty unlikely case.  Reusing the same type parameter name is highly discouraged in C# (the compiler even reports a warning for this).  As such, most people will gives these non conflicting names, and the change will be fine.

If this is felt to be important enough to fix, it will take medium effort to fix.  Specifically, the analysis code will have to see which type parameters were captured in the parameter and return type of the lambda.  It will then have to see if, from the replacement location, the names of those type parameters bind back to that type parameter and not to some other type in scope (like another nested type, or a method/class type parameter, etc.).

**Comment by CyrusNajmabadi** [comment]
@chborl Kevin asked me to look at https://github.com/dotnet/roslyn/issues/23118.  IN the process of doing so, i ended up addressing most of the cases of https://github.com/dotnet/roslyn/issues/22672 since it was all in the exact same code location.  I hope that's ok!

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -98,14 +109,14 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
-            if (IsWrittenAfter(semanticModel, local, block, anonymousFunction, cancellationToken))
+            var delegateType = semanticModel.GetTypeInfo(anonymousFunction, cancellationToken).ConvertedType as INamedTypeSymbol;
+            if (!delegateType.IsDelegateType() ||
+                delegateType.DelegateInvokeMethod == null)
             {
                 return;
             }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: this code was already here.  i just swapped it with the complex check which now comes afterwards.

**Comment by sharwell** [comment]
> If this is felt to be important enough to fix

Let's just file this as a follow-up bug.

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -203,15 +216,49 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
+
+                        if (nodeToCheck.Parent is MemberAccessExpressionSyntax)
+                        {
+                            // They're doing something like "del.ToString()".  Can't do this with a
+                            // local function.
+                            return false;
+                        }
+
+                        var convertedType = semanticModel.GetTypeInfo(nodeToCheck, cancellationToken).ConvertedType;
+                        if (convertedType == null || !convertedType.IsDelegateType())
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: :bulb: `IsDelegateType()` is an extension method which handles `null` internally. The null check here is unnecessary.

File: src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs [hunk 3]
```diff
@@ -1191,5 +1191,295 @@ void M()
...
+{
+  delegate T MyDelegate(T t = null);
+
+  public class Class
+  {
+    public void Caller()
+    {
+      MyDelegate [||]local = x => x;
+
+      local.Invoke();
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UseLocalFunction/UseLocalFunctionTests.cs]: :bulb: Would be good to file a follow-up issue (feature request) to handle the case where the only member access is `Invoke`.

**Comment by sharwell** [comment]
@CyrusNajmabadi Let me know when you have seen the feedback above and I'll move it forward after you are done with any changes you wanted to make.

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: Done.

**Comment by CyrusNajmabadi** [comment]
Have made the code change.  Feel free to do what you want now :)

**Comment by chborl** [comment]
Thanks, @CyrusNajmabadi  :)

**Comment by CyrusNajmabadi** [comment]
Any time!

**Pull request merged**